src/query/sql/src/planner/optimizer/property/enforcer.rs

```diff
@@ -57,12 +57,19 @@ pub fn require_property(
         }
         if let RelOperator::UnionAll(_) = &s_expr.plan {
             // Wrap the child with Random exchange to make it partition to all nodes
-            let enforced_child =
-                enforce_property(optimized_expr.child(index)?, &RequiredProperty {
-                    distribution: Distribution::Any,
-                })?;
-            children.push(enforced_child);
-            continue;
+            // Check if exists `Merge` in child, if not exits, wrap it with `Exchange`
+            if optimized_expr
+                .children()
+                .iter()
+                .all(|child| !check_merge(child))
+            {
+                let enforced_child =
+                    enforce_property(optimized_expr.child(index)?, &RequiredProperty {
+                        distribution: Distribution::Random,
```

Comment:
`Distribution::Random` is used to describe derived property and shouldn't be used in `RequiredProperty`.